src/Compiler/Checking/CheckDeclarations.fs

```diff
@@ -3162,8 +3162,21 @@ module EstablishTypeDefinitionCores =
                     if not ctorArgNames.IsEmpty then errorR (Error(FSComp.SR.parsOnlyClassCanTakeValueArguments(), m))
                 
             let envinner = AddDeclaredTypars CheckForDuplicateTypars (tycon.Typars m) envinner
-            let envinner = MakeInnerEnvForTyconRef envinner thisTyconRef false 
+            let envinner = MakeInnerEnvForTyconRef envinner thisTyconRef false
+            
+            let multiCaseUnionStructCheck (unionCases: UnionCase list) =
+                if tycon.IsStructRecordOrUnionTycon && unionCases.Length > 1 then 
+                    let fieldNames = [ for uc in unionCases do for ft in uc.FieldTable.TrueInstanceFieldsAsList do yield ft ]
+                    let fieldNames = fieldNames |> List.map(fun x -> (x.LogicalName, x.rfield_name_generated, x.Range))
+
+                    if fieldNames |> List.distinctBy(fun (name, _, _) -> name) |> List.length <> fieldNames.Length then
+                        let fieldRanges =
+                            fieldNames
+                            |> List.filter(fun (name, isGenerated, _) -> (not isGenerated || name = "Item") || isGenerated)
```

Comment:
This will always be `true` so it's either redundant or needs to be fixed.

Reply:
@0101 
This is to cover this scenario. 
-  | A of Item: int  -> Item is not generated(manually added by user)
-  | B of string  -> compiler autogenerates Item as a LogicalName so we need to report it also


```fsharp

     [<Fact>]
     let ``If a union type has more than one case and is a struct, field must be given unique name 15`` () =
         Fsx """
 namespace Foo
 [<Struct>]
 type StructUnion =
     | A of Item: int
     | B of string
         """
         |> compile
         |> shouldFail
         |> withDiagnostics [
             (Error 3204, Line 5, Col 12, Line 5, Col 16, "If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).")
             (Error 3204, Line 6, Col 12, Line 6, Col 18, "If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).")
         ]
```

Reply:
@edgarfgp in that case you probably want:
```fsharp
(not isGenerated && name = "Item") || isGenerated
```
or simpler:
```fsharp
isGenerated || name = "Item"
```

But since all the tests passed anyway - what would be the case which this is supposed to filter out?

Reply:
Yeah, I think I was trying to be too clever :) and show only the error on ranges that not auto generated. But I think it is safer to remove the filter and just show the error on the ranges that are in the list.

Reply:
@0101 remove the need of filtering. Thanks for the review